File: src/Python/05_scheduling.py
#!/usr/bin/env python3

from scripts.communication_helpers.communication_helper import Command, send_message

import time
import sched
import datetime

scheduler = sched.scheduler(time.time, time.sleep)
start_time = ''

def function_1():
  if datetime.datetime.now() - start_time < datetime.timedelta(seconds=10):
    scheduler.enter(1, 0, function_1, ())

if __name__ == "__main__":
  send_message(Command.START, 'scheduling')

  start_time = datetime.datetime.now()
  scheduler.enter(0, 0, function_1, ())
  scheduler.run()

  send_message(Command.STOP, 'scheduling')

File: src/C++/05_scheduling.cpp
/*
 * Compiling:
 * g++ 05_scheduling.cpp                                       \
 *   ../../scripts/communication_helpers/CommunicationHelper.h \
 *   ../../scripts/communication_helpers/CommunicationHelper.c \
 *   -lwiringPi
 */

#include <ctime>
#include <chrono>
#include <iostream>
#include <thread>

#include "../../scripts/communication_helpers/CommunicationHelper.h"

using namespace std;
using namespace std::chrono;

void func() {
  ;
}

void start(int interval, function<void(void)> func) {
  std::thread([interval, func]() {
    while (true) {
      func();
      this_thread::sleep_for(milliseconds(interval));
    }
  }).detach();
}

int main () {

  SendCommand(START, "scheduling_cpp");

  int timeout = 10;
  auto startTime = system_clock::now();

  start(1000, &func);

  duration<double> ellapsedTime;
  do {
    ellapsedTime = system_clock::now() - startTime;
  } while (ellapsedTime.count() < timeout);

  SendCommand(STOP, "scheduling_cpp");

  return 0;
}